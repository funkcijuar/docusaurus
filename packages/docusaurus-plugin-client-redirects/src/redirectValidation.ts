/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {isValidPathname} from '@docusaurus/utils';
import * as Yup from 'yup';
import {RedirectMetadata} from './types';

const validPathnameTest: Yup.TestOptions = {
  name: 'isValidPathname',
  message:
    '${path} is not a valid pathname. Pathname should start with / and not contain any domain or query string',
  test: isValidPathname,
};

const RedirectSchema = Yup.object<RedirectMetadata>({
  fromRoutePath: Yup.string().required().test(validPathnameTest),
  toRoutePath: Yup.string().required().test(validPathnameTest),
});

export function validateRedirect(redirect: RedirectMetadata) {
  RedirectSchema.validateSync(redirect);
}
