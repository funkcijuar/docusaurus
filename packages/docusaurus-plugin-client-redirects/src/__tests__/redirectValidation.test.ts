/**
 * Copyright (c) 2017-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {validateRedirect} from '../redirectValidation';

describe('validateRedirect', () => {
  test('validate good redirects without throwing', () => {
    validateRedirect({
      fromRoutePath: '/fromSomePath',
      toRoutePath: '/toSomePath',
    });
    validateRedirect({
      fromRoutePath: '/from/Some/Path',
      toRoutePath: '/toSomePath',
    });
    validateRedirect({
      fromRoutePath: '/fromSomePath',
      toRoutePath: '/toSomePath',
    });
    validateRedirect({
      fromRoutePath: '/fromSomePath',
      toRoutePath: '/to/Some/Path',
    });
  });

  test('throw for bad redirects', () => {
    expect(() =>
      validateRedirect({
        fromRoutePath: 'https://fb.com/fromSomePath',
        toRoutePath: '/toSomePath',
      }),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      validateRedirect({
        fromRoutePath: '/fromSomePath',
        toRoutePath: 'https://fb.com/toSomePath',
      }),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      validateRedirect({
        fromRoutePath: '/fromSomePath',
        toRoutePath: '/toSomePath?queryString=xyz',
      }),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      validateRedirect({
        fromRoutePath: null as any,
        toRoutePath: '/toSomePath?queryString=xyz',
      }),
    ).toThrowErrorMatchingSnapshot();

    expect(() =>
      validateRedirect({
        fromRoutePath: ['heyho'] as any,
        toRoutePath: '/toSomePath?queryString=xyz',
      }),
    ).toThrowErrorMatchingSnapshot();
  });
});
