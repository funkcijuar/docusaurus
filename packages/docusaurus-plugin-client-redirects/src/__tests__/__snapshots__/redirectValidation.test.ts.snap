// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateRedirect throw for bad redirects 1`] = `
"fromRoutePath is not a valid pathname. Pathname should start with / and not contain any domain or query string
Redirect={\\"fromRoutePath\\":\\"https://fb.com/fromSomePath\\",\\"toRoutePath\\":\\"/toSomePath\\"}"
`;

exports[`validateRedirect throw for bad redirects 2`] = `
"toRoutePath is not a valid pathname. Pathname should start with / and not contain any domain or query string
Redirect={\\"fromRoutePath\\":\\"/fromSomePath\\",\\"toRoutePath\\":\\"https://fb.com/toSomePath\\"}"
`;

exports[`validateRedirect throw for bad redirects 3`] = `
"toRoutePath is not a valid pathname. Pathname should start with / and not contain any domain or query string
Redirect={\\"fromRoutePath\\":\\"/fromSomePath\\",\\"toRoutePath\\":\\"/toSomePath?queryString=xyz\\"}"
`;
